File: bench/corpus_crossval.zig
const bcs = @import("bcs");
const std = @import("std");

const corpus = @embedFile("malformed_corpus.txt");

const MutationEnum = union(enum) {
    Unit,
    U64: u64,
    Bytes: [4]u8,
    Pair: struct { a: u16, b: u8 },
};

const Nested = struct {
    id: u16,
    names: []const bcs.String,
    flag: bool,
};

fn errName(err: bcs.Error) []const u8 {
    return switch (err) {
        error.UnexpectedEndOfInput => "eof",
        error.SequenceTooLong => "sequence_too_long",
        error.ContainerTooDeep => "container_too_deep",
        error.InvalidBool => "invalid_bool",
        error.NonCanonicalMap => "non_canonical_map",
        error.InvalidOptionTag => "invalid_option_tag",
        error.TrailingBytes => "trailing_bytes",
        error.Utf8 => "utf8",
        error.NonCanonicalUleb128 => "non_canonical_uleb128",
        error.Uleb128Overflow => "uleb128_overflow",
        error.NotSupported => "not_supported",
        error.InvalidEnumTag => "invalid_enum_tag",
        error.OutOfMemory => "out_of_memory",
    };
}

fn emitStatus(comptime T: type, name: []const u8, raw: []const u8) void {
    const allocator = std.heap.c_allocator;
    const value = bcs.deserialize(T, allocator, raw) catch |err| {
        std.debug.print("{s}=err:{s}\n", .{ name, errName(err) });
        return;
    };
    defer bcs.freeDeserialized(T, allocator, value);

    const encoded = bcs.serialize(allocator, value) catch |err| {
        std.debug.print("{s}=err:{s}\n", .{ name, errName(err) });
        return;
    };
    defer allocator.free(encoded);

    std.debug.print("{s}=ok:", .{name});
    for (encoded) |byte| std.debug.print("{x:0>2}", .{byte});
    std.debug.print("\n", .{});
}

fn decodeHex(allocator: std.mem.Allocator, hex: []const u8) ![]u8 {
    if (hex.len % 2 != 0) return error.InvalidHex;

    const bytes = try allocator.alloc(u8, hex.len / 2);
    errdefer allocator.free(bytes);

    var i: usize = 0;
    while (i < bytes.len) : (i += 1) {
        const hi = try std.fmt.charToDigit(hex[i * 2], 16);
        const lo = try std.fmt.charToDigit(hex[i * 2 + 1], 16);
        bytes[i] = @as(u8, @intCast((hi << 4) | lo));
    }
    return bytes;
}

fn dispatch(name: []const u8, ty: []const u8, raw: []const u8) void {
    if (std.mem.eql(u8, ty, "bool")) {
        emitStatus(bool, name, raw);
    } else if (std.mem.eql(u8, ty, "opt_u8")) {
        emitStatus(?u8, name, raw);
    } else if (std.mem.eql(u8, ty, "string")) {
        emitStatus(bcs.String, name, raw);
    } else if (std.mem.eql(u8, ty, "bytes")) {
        emitStatus([]const u8, name, raw);
    } else if (std.mem.eql(u8, ty, "vec_u16")) {
        emitStatus([]const u16, name, raw);
    } else if (std.mem.eql(u8, ty, "tuple_u32_bool")) {
        emitStatus(struct { u32, bool }, name, raw);
    } else if (std.mem.eql(u8, ty, "mut_enum")) {
        emitStatus(MutationEnum, name, raw);
    } else if (std.mem.eql(u8, ty, "nested")) {
        emitStatus(Nested, name, raw);
    } else if (std.mem.eql(u8, ty, "map_u8_unit")) {
        emitStatus(bcs.Map(u8, void), name, raw);
    } else if (std.mem.eql(u8, ty, "map_string_u32")) {
        emitStatus(bcs.Map(bcs.String, u32), name, raw);
    } else {
        unreachable;
    }
}

pub fn main() !void {
    const allocator = std.heap.c_allocator;
    var lines = std.mem.splitScalar(u8, corpus, '\n');
    while (lines.next()) |line_raw| {
        const line = std.mem.trim(u8, line_raw, " \r\t");
        if (line.len == 0 or line[0] == '#') continue;

        var parts = std.mem.splitScalar(u8, line, '|');
        const name = parts.next() orelse return error.InvalidCorpus;
        const ty = parts.next() orelse return error.InvalidCorpus;
        const hex = parts.next() orelse return error.InvalidCorpus;
        if (parts.next() != null) return error.InvalidCorpus;

        const bytes = try decodeHex(allocator, hex);
        defer allocator.free(bytes);

        dispatch(name, ty, bytes);
    }
}

File: bench/corpus_crossval.rs
use serde::{de::DeserializeOwned, Deserialize, Serialize};
use std::collections::BTreeMap;

const CORPUS: &str = include_str!("malformed_corpus.txt");

#[derive(Clone, Debug, Serialize, Deserialize)]
enum MutationEnum {
    Unit,
    U64(u64),
    Bytes([u8; 4]),
    Pair { a: u16, b: u8 },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
struct Nested {
    id: u16,
    names: Vec<String>,
    flag: bool,
}

fn hex(bytes: &[u8]) -> String {
    bytes
        .iter()
        .map(|b| format!("{:02x}", b))
        .collect::<Vec<_>>()
        .join("")
}

fn err_name(err: &bcs::Error) -> &'static str {
    match err {
        bcs::Error::Eof
        | bcs::Error::Io(_)
        | bcs::Error::ExpectedMapKey
        | bcs::Error::ExpectedMapValue
        | bcs::Error::MissingLen => "eof",
        bcs::Error::ExceededMaxLen(_) => "sequence_too_long",
        bcs::Error::ExceededContainerDepthLimit(_) => "container_too_deep",
        bcs::Error::ExpectedBoolean => "invalid_bool",
        bcs::Error::NonCanonicalMap => "non_canonical_map",
        bcs::Error::ExpectedOption => "invalid_option_tag",
        bcs::Error::RemainingInput => "trailing_bytes",
        bcs::Error::Utf8 => "utf8",
        bcs::Error::NonCanonicalUleb128Encoding => "non_canonical_uleb128",
        bcs::Error::IntegerOverflowDuringUleb128Decoding => "uleb128_overflow",
        bcs::Error::NotSupported(_) => "not_supported",
        bcs::Error::Custom(msg) if msg.contains("expected variant index") => "invalid_enum_tag",
        bcs::Error::Custom(_) => "custom",
    }
}

fn emit_status<T>(name: &str, bytes: &[u8])
where
    T: Serialize + DeserializeOwned,
{
    match bcs::from_bytes::<T>(bytes) {
        Ok(value) => {
            let encoded = bcs::to_bytes(&value).unwrap();
            println!("{name}=ok:{}", hex(&encoded));
        }
        Err(err) => println!("{name}=err:{}", err_name(&err)),
    }
}

fn decode_hex(hex: &str) -> Vec<u8> {
    assert!(hex.len() % 2 == 0, "invalid hex length for {hex}");
    let mut bytes = Vec::with_capacity(hex.len() / 2);
    let chars = hex.as_bytes();
    let mut i = 0;
    while i < chars.len() {
        let byte = u8::from_str_radix(&hex[i..i + 2], 16).unwrap();
        bytes.push(byte);
        i += 2;
    }
    bytes
}

fn dispatch(name: &str, ty: &str, bytes: &[u8]) {
    match ty {
        "bool" => emit_status::<bool>(name, bytes),
        "opt_u8" => emit_status::<Option<u8>>(name, bytes),
        "string" => emit_status::<String>(name, bytes),
        "bytes" => emit_status::<Vec<u8>>(name, bytes),
        "vec_u16" => emit_status::<Vec<u16>>(name, bytes),
        "tuple_u32_bool" => emit_status::<(u32, bool)>(name, bytes),
        "mut_enum" => emit_status::<MutationEnum>(name, bytes),
        "nested" => emit_status::<Nested>(name, bytes),
        "map_u8_unit" => emit_status::<BTreeMap<u8, ()>>(name, bytes),
        "map_string_u32" => emit_status::<BTreeMap<String, u32>>(name, bytes),
        _ => panic!("unknown corpus type: {ty}"),
    }
}

fn main() {
    for line in CORPUS.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with('#') {
            continue;
        }

        let mut parts = line.split('|');
        let name = parts.next().unwrap();
        let ty = parts.next().unwrap();
        let hex = parts.next().unwrap_or("");
        assert!(parts.next().is_none(), "too many fields in corpus line: {line}");

        let bytes = decode_hex(hex);
        dispatch(name, ty, &bytes);
    }
}
